src/codec/lz4/encoder.rs

```diff
@@ -0,0 +1,309 @@
+        }
+
+        let min_dst_size = match &lz4_fn {
+            Lz4Fn::Begin => LZ4F_HEADER_SIZE_MAX,
+            Lz4Fn::Update { input } => {
+                min_dst_size(input.unwritten().len().min(self.limit), &self.preferences)
+            }
+            Lz4Fn::Flush | Lz4Fn::End => self.flush_buffer_size,
+        };
+
+        let direct_output = output.unwritten().len() >= min_dst_size;
+
+        let (dst_buffer, dst_max_size) = if direct_output {
+            let output_len = output.unwritten().len();
+            (output.unwritten_mut(), output_len)
+        } else {
+            let buffer_size = self.block_buffer_size;
+            let buffer = self
+                .maybe_buffer
+                .get_or_insert_with(|| PartialBuffer::new(Vec::with_capacity(buffer_size)));
+            buffer.reset();
+            (buffer.unwritten_mut(), buffer_size)
+        };
+
+        let len = match lz4_fn {
+            Lz4Fn::Begin => {
+                let len = check_error(unsafe {
+                    LZ4F_compressBegin(
+                        self.ctx.get_mut().ctx,
+                        dst_buffer.as_mut_ptr(),
+                        dst_max_size,
+                        &self.preferences,
+                    )
+                })?;
+                self.state = State::Encoding;
+                len
+            }
+            Lz4Fn::Update { input } => {
+                let size = input.unwritten().len().min(self.limit);
+                let len = check_error(unsafe {
+                    LZ4F_compressUpdate(
+                        self.ctx.get_mut().ctx,
+                        dst_buffer.as_mut_ptr(),
+                        dst_max_size,
+                        input.unwritten().as_ptr(),
+                        size,
+                        core::ptr::null(),
+                    )
+                })?;
+                input.advance(size);
+                len
+            }
+            Lz4Fn::Flush => check_error(unsafe {
+                LZ4F_flush(
+                    self.ctx.get_mut().ctx,
+                    dst_buffer.as_mut_ptr(),
+                    dst_max_size,
+                    core::ptr::null(),
+                )
+            })?,
+            Lz4Fn::End => {
+                let len = check_error(unsafe {
+                    LZ4F_compressEnd(
+                        self.ctx.get_mut().ctx,
+                        dst_buffer.as_mut_ptr(),
+                        dst_max_size,
+                        core::ptr::null(),
+                    )
+                })?;
+                self.state = State::Footer;
+                len
+            }
+        };
+
+        let drained_after = if direct_output {
+            output.advance(len);
+            len
+        } else {
+            // SAFETY: buffer is initialized above incase of a non-direct operation
+            unsafe {
+                self.maybe_buffer
+                    .as_mut()
+                    .unwrap_unchecked()
```

Comment:
I'm slightly worried about this `unwrap_unchecked`, would be good if we can optimize this out the `unwrap` completely, i.e. by returning another option along side `dst_buffer`